File: prerequest.js
const fs = require('fs');
require('dotenv').config();
const crypto = require('crypto');
const path = require('path');

const envPath = path.resolve(process.cwd(), '.env');

// Function to append environment variable to .env file
const appendToEnvFile = (key, value) => {
    const envVarString = `${key}=${value}`;
    // Check if the key already exists and replace it
    let envContent = fs.existsSync(envPath) ? fs.readFileSync(envPath, 'utf-8') : '';
    const regex = new RegExp(`^${key}=.*`, 'gm');
    if (envContent.match(regex)) {
        envContent = envContent.replace(regex, envVarString);
    } else {
        envContent += envVarString;
    }
    fs.writeFileSync(envPath, envContent, 'utf-8');
};


// Convert variables (You might not need this function if you are directly using environment variables)
const getEnvVar = (variable) => process.env[variable];

// Unique Key generation based on the current date
const uniqueKey = (date) => {
    const nonce = date.toISOString().slice(0, 19) + '.000Z';
    return getEnvVar('BSP_SECRET_KEY') + nonce;
};

// Signable content creation
const signableContent = (method, url, contentType, contentMd5, organization) => {
    const requestPath = url.replace(/^https?:\/\/[^\/]+\//, '/');
    return [method, requestPath, contentType, contentMd5, organization].filter(p => p && p.length > 0).join('\n');
};

// Adjust calculateSignature to return both date and signature
const calculateSignature = () => {
    const date = new Date().toGMTString(); // Store the GMT string format of the date
    const key = uniqueKey(new Date(date));
    const organization = process.env.BSP_ORGANIZATION;
    // Example request details
    const method = 'POST';
    const url = "/order/3/orders/1";
    const contentType = 'application/json';
    const contentMd5 = undefined;

    const sc = signableContent(method, url, contentType, contentMd5, organization);
    const hmac = crypto.createHmac('sha512', key).update(sc).digest('base64');
    return { date, signature: hmac };
};

// Update generateAccessKey to use the modified calculateSignature
const generateAccessKey = () => {
    const { date, signature } = calculateSignature();
    const sharedKey = process.env.BSP_SHARED_KEY;
    const accessKey = `AccessKey ${sharedKey}:${signature}`;
    console.log('Generated BSP Access Key:', accessKey);

    // Append DATE and BSP_ACCESS_KEY to .env file
    appendToEnvFile('DATE', date);
    appendToEnvFile('BSP_ACCESS_KEY', accessKey);
//     return {accessKey, date};
};

generateAccessKey();

File: drive_thru_bot/pos/prerequest.py
import os
import datetime
import hashlib
import base64
from pytz import timezone  # type: ignore


# Convert variables
def get_env_var(variable):
    return os.getenv(variable)


# Unique Key generation based on the current date
def unique_key(date):
    nonce = date.astimezone(timezone("UTC")).isoformat(timespec="seconds") + ".000Z"
    return get_env_var("BSP_SECRET_KEY") + nonce


# Signable content creation
def signable_content(method, url, content_type, content_md5, organization):
    request_path = url.replace(r"^https?:\/\/[^\/]+\/", "/", 1)
    return "\n".join(
        filter(None, [method, request_path, content_type, content_md5, organization])
    )


# Adjust calculateSignature to return both date and signature
def calculate_signature():
    date = datetime.datetime.now(datetime.timezone.utc).strftime(
        "%a, %d %b %Y %H:%M:%S GMT"
    )
    key = unique_key(datetime.datetime.now(datetime.timezone.utc))
    organization = get_env_var("BSP_ORGANIZATION")
    # Example request details
    method = "POST"
    url = "/order/3/orders/1"
    content_type = "application/json"
    content_md5 = None

    sc = signable_content(method, url, content_type, content_md5, organization)
    signature = base64.b64encode(
        hashlib.sha512(sc.encode("utf-8") + key.encode("utf-8")).digest()
    ).decode("utf-8")
    return {"date": date, "signature": signature}


# Update generateAccessKey to use the modified calculateSignature
def generate_date_and_access_key():
    result = calculate_signature()
    date = result["date"]
    signature = result["signature"]
    shared_key = get_env_var("BSP_SHARED_KEY")
    access_key = f"AccessKey {shared_key}:{signature}"
    print("Generated BSP Access Key:", access_key)

    # For appending to .env file, you'd need to handle file operations directly in Python
    # This example does not include the file append logic directly but it's straightforward with open()

    return access_key, date


# This example assumes the existence of an environment with proper variables and timezone handling with pytz.
# Adjustments might be needed based on your specific environment and requirements.
